README.md
# Spotifier

Simple Spotify search library that provides song search and best match retrieval from the Spotify API

## Install

```bash
npm install spotifier
```

## Usage:

In your code, require the `spotifier` module as follows:

### Artist and Track search

```javascript
var spotifier = require('spotifier');

var
  client = spotifier(),
  params = {
    artist : 'foo fighters',
    title : 'breakout'
  };

client.search(params, function (err, result) {
  // work with results here...
});
```

### ISRC search

```javascript
var spotifier = require('spotifier');

var
  client = spotifier(),
  params = {
    isrc : 'USWB10303069'
  };

client.search(params, function (err, result) {
  // work with results here...
});
```

### Options:

* authorizationUrl - URL of authorization endpoint (defaults to `https://accounts.spotify.com/api/token`)
* clientId - Optional - when specified, client credentials are authenticated and the resulting access token is used for subsequent search requests
* clientSecret - Optional - when specified, client credentials are authenticated and the resulting access token is used for subsequent search requests
* market - Optional - when specified, searches for matching tracks will utilize the specified market
* searchResultLimit - Limit of results to return (defaults to 10)
* searchUrl - URL of API endpoint (defaults to `https://api.spotify.com/v1/search`)
* timeout - Timeout for request in milliseconds (defaults to `10000`)

```javascript
var spotifier = require('spotifier');

var client = spotifier({
  albumUrl : 'https://api.spotify.com/v1/albums',
  authorizationUrl : 'https://accounts.spotify.com/api/token',
  clientId : '<client_id value here>',
  clientSecret : '<client_secret value here>',
  searchResultLimit : 10,
  searchUrl : 'https://api.spotify.com/v1/search',
  timeout : 10000
});
```

### Find Best Match

Find the best match (or matches) for a search:

`client.findBestMatch(options, params, callback)`

Where options is an _optional_ object supporting the following fields:

+ `multi` _(optional, boolean)_ - defaults to `false`, when `true` all the closest matches are returned as an array

Where params is an object supporting the following fields:

+ `isrc` _(optional, string)_ - when specified, performs an ISRC search with Spotify
+ `artist` _(optional, string)_ - when specified (instead of `isrc`) performs an artist and title search with Spotify
+ `title` _(optional, string)_ - when specified (instead of `isrc`) performs an artist and title search with Spotify
+ `market` _(optional, string)_ - when specified, performs a search with with the specified market

```javascript
var
  client = (require('spotifier')()),
  options = {
    multi : false
  },
  params = {
    artist : 'foo fighters',
    title : 'breakout'
  };

client.findBestMatch(options, params, function (err, result) {
  if (err) {
    console.error(err);
  }

  console.log(result);
});
```

Which returns the following response:

```javascript
{ album:
   { album_type: 'album',
     available_markets:
      [ 'AD',
        'AR',
        'AT',
        'AU',
        'BE',
        'BG',
        'BO',
        'BR',
        'CH',
        'CL',
        'CO',
        'CR',
        'CY',
        'CZ',
        'DE',
        'DK',
        'DO',
        'EC',
        'EE',
        'ES',
        'FI',
        'FR',
        'GB',
        'GR',
        'GT',
        'HK',
        'HN',
        'HU',
        'IE',
        'IS',
        'IT',
        'LI',
        'LT',
        'LU',
        'LV',
        'MC',
        'MY',
        'NI',
        'NL',
        'NO',
        'NZ',
        'PA',
        'PE',
        'PH',
        'PL',
        'PT',
        'PY',
        'RO',
        'SE',
        'SG',
        'SI',
        'SK',
        'SV',
        'TR',
        'TW',
        'UY' ],
     external_urls: { spotify: 'https://open.spotify.com/album/7MKi8PuCX5QJtaG9wLbOI2' },
     href: 'https://api.spotify.com/v1/albums/7MKi8PuCX5QJtaG9wLbOI2',
     id: '7MKi8PuCX5QJtaG9wLbOI2',
     images: [ [Object], [Object], [Object] ],
     name: 'Mellow Gold',
     type: 'album',
     uri: 'spotify:album:7MKi8PuCX5QJtaG9wLbOI2' },
  artists:
   [ { external_urls: [Object],
       href: 'https://api.spotify.com/v1/artists/3vbKDsSS70ZX9D2OcvbZmS',
       id: '3vbKDsSS70ZX9D2OcvbZmS',
       name: 'Beck',
       type: 'artist',
       uri: 'spotify:artist:3vbKDsSS70ZX9D2OcvbZmS' } ],
  available_markets:
   [ 'AD',
     'AR',
     'AT',
     'AU',
     'BE',
     'BG',
     'BO',
     'BR',
     'CH',
     'CL',
     'CO',
     'CR',
     'CY',
     'CZ',
     'DE',
     'DK',
     'DO',
     'EC',
     'EE',
     'ES',
     'FI',
     'FR',
     'GB',
     'GR',
     'GT',
     'HK',
     'HN',
     'HU',
     'IE',
     'IS',
     'IT',
     'LI',
     'LT',
     'LU',
     'LV',
     'MC',
     'MY',
     'NI',
     'NL',
     'NO',
     'NZ',
     'PA',
     'PE',
     'PH',
     'PL',
     'PT',
     'PY',
     'RO',
     'SE',
     'SG',
     'SI',
     'SK',
     'SV',
     'TR',
     'TW',
     'UY' ],
  disc_number: 1,
  duration_ms: 234960,
  explicit: false,
  external_ids: { isrc: 'USGF19463401' },
  external_urls: { spotify: 'https://open.spotify.com/track/4b8GTae6rqb9oPiRzVa3Gy' },
  href: 'https://api.spotify.com/v1/tracks/4b8GTae6rqb9oPiRzVa3Gy',
  id: '4b8GTae6rqb9oPiRzVa3Gy',
  name: 'Loser',
  popularity: 67,
  preview_url: 'https://p.scdn.co/mp3-preview/5efa81cf9d4d2bc26407e928a0a150a7bc4ecc66',
  track_number: 1,
  type: 'track',
  uri: 'spotify:track:4b8GTae6rqb9oPiRzVa3Gy' }
```

### Search

Find all matches for a search:

`client.search(options, params, callback)`

_Please note, the `options` parameter is optional and can be omitted_

```javascript
var
  client = (require('spotifier')()),
  options = {
    limit : 2,
    start : 0
  },
  params = {
    artist : 'beck',
    title : 'loser'
  };

client.search(options, params, function (err, result) {
  if (err) {
    console.error(err);
  }

  console.log(result);
});
```

Which returns the following response:

```javascript
{
  "href": "https://api.spotify.com/v1/search?query=artist%3Abeck+track%3Aloser&offset=0&limit=2&type=track",
  "items": [
    {
      "album": {
        "album_type": "album",
        "available_markets": [
          "AD",
          "AR",
          "AT",
          "AU",
          "BE",
          "BG",
          "BO",
          "BR",
          "CH",
          "CL",
          "CO",
          "CR",
          "CY",
          "CZ",
          "DE",
          "DK",
          "DO",
          "EC",
          "EE",
          "ES",
          "FI",
          "FR",
          "GB",
          "GR",
          "GT",
          "HK",
          "HN",
          "HU",
          "IE",
          "IS",
          "IT",
          "LI",
          "LT",
          "LU",
          "LV",
          "MC",
          "MY",
          "NI",
          "NL",
          "NO",
          "NZ",
          "PA",
          "PE",
          "PH",
          "PL",
          "PT",
          "PY",
          "RO",
          "SE",
          "SG",
          "SI",
          "SK",
          "SV",
          "TR",
          "TW",
          "UY"
        ],
        "external_urls": {
          "spotify": "https://open.spotify.com/album/7MKi8PuCX5QJtaG9wLbOI2"
        },
        "href": "https://api.spotify.com/v1/albums/7MKi8PuCX5QJtaG9wLbOI2",
        "id": "7MKi8PuCX5QJtaG9wLbOI2",
        "images": [
          {
            "height": 638,
            "url": "https://i.scdn.co/image/2f4e58958061664dbe1666936dab319c70398d8b",
            "width": 640
          },
          {
            "height": 299,
            "url": "https://i.scdn.co/image/0f057142f11c251f81a22ca639b7261530b280b2",
            "width": 300
          },
          {
            "height": 64,
            "url": "https://i.scdn.co/image/03e3cfd4c4de65a906dfa50a7e11203994935639",
            "width": 64
          }
        ],
        "name": "Mellow Gold",
        "type": "album",
        "uri": "spotify:album:7MKi8PuCX5QJtaG9wLbOI2"
      },
      "artists": [
        {
          "external_urls": {
            "spotify": "https://open.spotify.com/artist/3vbKDsSS70ZX9D2OcvbZmS"
          },
          "href": "https://api.spotify.com/v1/artists/3vbKDsSS70ZX9D2OcvbZmS",
          "id": "3vbKDsSS70ZX9D2OcvbZmS",
          "name": "Beck",
          "type": "artist",
          "uri": "spotify:artist:3vbKDsSS70ZX9D2OcvbZmS"
        }
      ],
      "available_markets": [
        "AD",
        "AR",
        "AT",
        "AU",
        "BE",
        "BG",
        "BO",
        "BR",
        "CH",
        "CL",
        "CO",
        "CR",
        "CY",
        "CZ",
        "DE",
        "DK",
        "DO",
        "EC",
        "EE",
        "ES",
        "FI",
        "FR",
        "GB",
        "GR",
        "GT",
        "HK",
        "HN",
        "HU",
        "IE",
        "IS",
        "IT",
        "LI",
        "LT",
        "LU",
        "LV",
        "MC",
        "MY",
        "NI",
        "NL",
        "NO",
        "NZ",
        "PA",
        "PE",
        "PH",
        "PL",
        "PT",
        "PY",
        "RO",
        "SE",
        "SG",
        "SI",
        "SK",
        "SV",
        "TR",
        "TW",
        "UY"
      ],
      "disc_number": 1,
      "duration_ms": 234960,
      "explicit": false,
      "external_ids": {
        "isrc": "USGF19463401"
      },
      "external_urls": {
        "spotify": "https://open.spotify.com/track/4b8GTae6rqb9oPiRzVa3Gy"
      },
      "href": "https://api.spotify.com/v1/tracks/4b8GTae6rqb9oPiRzVa3Gy",
      "id": "4b8GTae6rqb9oPiRzVa3Gy",
      "name": "Loser",
      "popularity": 67,
      "preview_url": "https://p.scdn.co/mp3-preview/5efa81cf9d4d2bc26407e928a0a150a7bc4ecc66",
      "track_number": 1,
      "type": "track",
      "uri": "spotify:track:4b8GTae6rqb9oPiRzVa3Gy"
    },
    {
      "album": {
        "album_type": "album",
        "available_markets": [
          "DE",
          "NO"
        ],
        "external_urls": {
          "spotify": "https://open.spotify.com/album/5fS7PzCkR0JLMKpNbvs1s9"
        },
        "href": "https://api.spotify.com/v1/albums/5fS7PzCkR0JLMKpNbvs1s9",
        "id": "5fS7PzCkR0JLMKpNbvs1s9",
        "images": [
          {
            "height": 640,
            "url": "https://i.scdn.co/image/b483c13fc566fff9f3ce6b195899c55f25277e9f",
            "width": 640
          },
          {
            "height": 300,
            "url": "https://i.scdn.co/image/2f919a375df9494d3071a3fb504ff805815cf3e9",
            "width": 300
          },
          {
            "height": 64,
            "url": "https://i.scdn.co/image/a459491722d38f8c75f0390eb5d03ff85b8f6f73",
            "width": 64
          }
        ],
        "name": "Rock",
        "type": "album",
        "uri": "spotify:album:5fS7PzCkR0JLMKpNbvs1s9"
      },
      "artists": [
        {
          "external_urls": {
            "spotify": "https://open.spotify.com/artist/3vbKDsSS70ZX9D2OcvbZmS"
          },
          "href": "https://api.spotify.com/v1/artists/3vbKDsSS70ZX9D2OcvbZmS",
          "id": "3vbKDsSS70ZX9D2OcvbZmS",
          "name": "Beck",
          "type": "artist",
          "uri": "spotify:artist:3vbKDsSS70ZX9D2OcvbZmS"
        }
      ],
      "available_markets": [
        "DE",
        "NO"
      ],
      "disc_number": 1,
      "duration_ms": 235706,
      "explicit": false,
      "external_ids": {
        "isrc": "USGF19463401"
      },
      "external_urls": {
        "spotify": "https://open.spotify.com/track/2m9cDUawCW5SfnpwHWtJJi"
      },
      "href": "https://api.spotify.com/v1/tracks/2m9cDUawCW5SfnpwHWtJJi",
      "id": "2m9cDUawCW5SfnpwHWtJJi",
      "name": "Loser",
      "popularity": 18,
      "preview_url": "https://p.scdn.co/mp3-preview/7ce39a2bd416792154386a9850ee813b25f21898",
      "track_number": 4,
      "type": "track",
      "uri": "spotify:track:2m9cDUawCW5SfnpwHWtJJi"
    }
  ],
  "limit": 2,
  "next": "https://api.spotify.com/v1/search?query=artist%3Abeck+track%3Aloser&offset=2&limit=2&type=track",
  "offset": 0,
  "previous": null,
  "total": 9
}
```

### Lookup Album

Retrieve a specific album from Spotify based on ID:

`client.lookupAlbum(options, albumId, callback)`

_Please note, the `options` parameter is optional and can be omitted_

```javascript
var client = (require('spotifier')());

client.search(options, '0lw68yx3MhKflWFqCsGkIs', function (err, result) {
  if (err) {
    console.error(err);
  }

  console.log(result);
});
```

Which returns the following response:

```javascript
{
  "album_type" : "album",
  "artists" : [ {
    "external_urls" : {
      "spotify" : "https://open.spotify.com/artist/12Chz98pHFMPJEknJQMWvI"
    },
    "href" : "https://api.spotify.com/v1/artists/12Chz98pHFMPJEknJQMWvI",
    "id" : "12Chz98pHFMPJEknJQMWvI",
    "name" : "Muse",
    "type" : "artist",
    "uri" : "spotify:artist:12Chz98pHFMPJEknJQMWvI"
  } ],
  "available_markets" : [ "AD", "AR", "AT", "AU", "BE", "BG", "BO", "BR", "CA", "CH", "CL", "CO", "CR", "CY", "CZ", "DE", "DK", "DO", "EC", "EE", "ES", "FI", "FR", "GB", "GR", "GT", "HK", "HN", "HU", "ID", "IE", "IS", "IT", "JP", "LI", "LT", "LU", "LV", "MC", "MT", "MX", "MY", "NI", "NL", "NO", "NZ", "PA", "PE", "PH", "PL", "PT", "PY", "SE", "SG", "SK", "SV", "TH", "TR", "TW", "US", "UY", "VN" ],
  "copyrights" : [ {
    "text" : "2006 A&E Records Limited",
    "type" : "C"
  }, {
    "text" : "2006 A&E Records Limited",
    "type" : "P"
  } ],
  "external_ids" : {
    "upc" : "825646911486"
  },
  "external_urls" : {
    "spotify" : "https://open.spotify.com/album/0lw68yx3MhKflWFqCsGkIs"
  },
  "genres" : [ ],
  "href" : "https://api.spotify.com/v1/albums/0lw68yx3MhKflWFqCsGkIs",
  "id" : "0lw68yx3MhKflWFqCsGkIs",
  "images" : [ {
    "height" : 640,
    "url" : "https://i.scdn.co/image/9e5288926fadb82f873ccf2b45300c3a6f65fa14",
    "width" : 640
  }, {
    "height" : 300,
    "url" : "https://i.scdn.co/image/f1cad0d6974d6236abd07a59106e8450d85cae24",
    "width" : 300
  }, {
    "height" : 64,
    "url" : "https://i.scdn.co/image/81a3f82578dc938c53efdcb405f6a3d3ebbf009f",
    "width" : 64
  } ],
  "label" : "Warner Bros.",
  "name" : "Black Holes And Revelations (Updated 09 version)",
  "popularity" : 75,
  "release_date" : "2006-06-19",
  "release_date_precision" : "day",
  "tracks" : {
    "href" : "https://api.spotify.com/v1/albums/0lw68yx3MhKflWFqCsGkIs/tracks?offset=0&limit=50",
    "items" : [ {
      "artists" : [ {
        "external_urls" : {
          "spotify" : "https://open.spotify.com/artist/12Chz98pHFMPJEknJQMWvI"
        },
        "href" : "https://api.spotify.com/v1/artists/12Chz98pHFMPJEknJQMWvI",
        "id" : "12Chz98pHFMPJEknJQMWvI",
        "name" : "Muse",
        "type" : "artist",
        "uri" : "spotify:artist:12Chz98pHFMPJEknJQMWvI"
      } ],
      "available_markets" : [ "AD", "AR", "AT", "AU", "BE", "BG", "BO", "BR", "CA", "CH", "CL", "CO", "CR", "CY", "CZ", "DE", "DK", "DO", "EC", "EE", "ES", "FI", "FR", "GB", "GR", "GT", "HK", "HN", "HU", "ID", "IE", "IS", "IT", "JP", "LI", "LT", "LU", "LV", "MC", "MT", "MX", "MY", "NI", "NL", "NO", "NZ", "PA", "PE", "PH", "PL", "PT", "PY", "SE", "SG", "SK", "SV", "TH", "TR", "TW", "US", "UY", "VN" ],
      "disc_number" : 1,
      "duration_ms" : 275493,
      "explicit" : false,
      "external_urls" : {
        "spotify" : "https://open.spotify.com/track/4jrCMOG9OPe6iF4vWFxatb"
      },
      "href" : "https://api.spotify.com/v1/tracks/4jrCMOG9OPe6iF4vWFxatb",
      "id" : "4jrCMOG9OPe6iF4vWFxatb",
      "name" : "Take a Bow",
      "preview_url" : "https://p.scdn.co/mp3-preview/4bf1c1a08cbf12759a37ba2b1f664b4b2cca9a31?cid=8897482848704f2a8f8d7c79726a70d4",
      "track_number" : 1,
      "type" : "track",
      "uri" : "spotify:track:4jrCMOG9OPe6iF4vWFxatb"
    }, {
      "artists" : [ {
        "external_urls" : {
          "spotify" : "https://open.spotify.com/artist/12Chz98pHFMPJEknJQMWvI"
        },
        "href" : "https://api.spotify.com/v1/artists/12Chz98pHFMPJEknJQMWvI",
        "id" : "12Chz98pHFMPJEknJQMWvI",
        "name" : "Muse",
        "type" : "artist",
        "uri" : "spotify:artist:12Chz98pHFMPJEknJQMWvI"
      } ],
      "available_markets" : [ "AD", "AR", "AT", "AU", "BE", "BG", "BO", "BR", "CA", "CH", "CL", "CO", "CR", "CY", "CZ", "DE", "DK", "DO", "EC", "EE", "ES", "FI", "FR", "GB", "GR", "GT", "HK", "HN", "HU", "ID", "IE", "IS", "IT", "JP", "LI", "LT", "LU", "LV", "MC", "MT", "MX", "MY", "NI", "NL", "NO", "NZ", "PA", "PE", "PH", "PL", "PT", "PY", "SE", "SG", "SK", "SV", "TH", "TR", "TW", "US", "UY", "VN" ],
      "disc_number" : 1,
      "duration_ms" : 240213,
      "explicit" : false,
      "external_urls" : {
        "spotify" : "https://open.spotify.com/track/3skn2lauGk7Dx6bVIt5DVj"
      },
      "href" : "https://api.spotify.com/v1/tracks/3skn2lauGk7Dx6bVIt5DVj",
      "id" : "3skn2lauGk7Dx6bVIt5DVj",
      "name" : "Starlight",
      "preview_url" : "https://p.scdn.co/mp3-preview/f7a1b8a270f310e43ced2720c9af5f29f6476b79?cid=8897482848704f2a8f8d7c79726a70d4",
      "track_number" : 2,
      "type" : "track",
      "uri" : "spotify:track:3skn2lauGk7Dx6bVIt5DVj"
    }, {
      "artists" : [ {
        "external_urls" : {
          "spotify" : "https://open.spotify.com/artist/12Chz98pHFMPJEknJQMWvI"
        },
        "href" : "https://api.spotify.com/v1/artists/12Chz98pHFMPJEknJQMWvI",
        "id" : "12Chz98pHFMPJEknJQMWvI",
        "name" : "Muse",
        "type" : "artist",
        "uri" : "spotify:artist:12Chz98pHFMPJEknJQMWvI"
      } ],
      "available_markets" : [ "AD", "AR", "AT", "AU", "BE", "BG", "BO", "BR", "CA", "CH", "CL", "CO", "CR", "CY", "CZ", "DE", "DK", "DO", "EC", "EE", "ES", "FI", "FR", "GB", "GR", "GT", "HK", "HN", "HU", "ID", "IE", "IS", "IT", "JP", "LI", "LT", "LU", "LV", "MC", "MT", "MX", "MY", "NI", "NL", "NO", "NZ", "PA", "PE", "PH", "PL", "PT", "PY", "SE", "SG", "SK", "SV", "TH", "TR", "TW", "US", "UY", "VN" ],
      "disc_number" : 1,
      "duration_ms" : 212439,
      "explicit" : false,
      "external_urls" : {
        "spotify" : "https://open.spotify.com/track/3lPr8ghNDBLc2uZovNyLs9"
      },
      "href" : "https://api.spotify.com/v1/tracks/3lPr8ghNDBLc2uZovNyLs9",
      "id" : "3lPr8ghNDBLc2uZovNyLs9",
      "name" : "Supermassive Black Hole",
      "preview_url" : "https://p.scdn.co/mp3-preview/7ab3e38ce1671da3a185d8685981983a6f39b7bd?cid=8897482848704f2a8f8d7c79726a70d4",
      "track_number" : 3,
      "type" : "track",
      "uri" : "spotify:track:3lPr8ghNDBLc2uZovNyLs9"
    }, {
      "artists" : [ {
        "external_urls" : {
          "spotify" : "https://open.spotify.com/artist/12Chz98pHFMPJEknJQMWvI"
        },
        "href" : "https://api.spotify.com/v1/artists/12Chz98pHFMPJEknJQMWvI",
        "id" : "12Chz98pHFMPJEknJQMWvI",
        "name" : "Muse",
        "type" : "artist",
        "uri" : "spotify:artist:12Chz98pHFMPJEknJQMWvI"
      } ],
      "available_markets" : [ "AD", "AR", "AT", "AU", "BE", "BG", "BO", "BR", "CA", "CH", "CL", "CO", "CR", "CY", "CZ", "DE", "DK", "DO", "EC", "EE", "ES", "FI", "FR", "GB", "GR", "GT", "HK", "HN", "HU", "ID", "IE", "IS", "IT", "JP", "LI", "LT", "LU", "LV", "MC", "MT", "MX", "MY", "NI", "NL", "NO", "NZ", "PA", "PE", "PH", "PL", "PT", "PY", "SE", "SG", "SK", "SV", "TH", "TR", "TW", "US", "UY", "VN" ],
      "disc_number" : 1,
      "duration_ms" : 258066,
      "explicit" : false,
      "external_urls" : {
        "spotify" : "https://open.spotify.com/track/5YXr4AGfUQpLSxtFSsKUh6"
      },
      "href" : "https://api.spotify.com/v1/tracks/5YXr4AGfUQpLSxtFSsKUh6",
      "id" : "5YXr4AGfUQpLSxtFSsKUh6",
      "name" : "Map Of The Problematique",
      "preview_url" : "https://p.scdn.co/mp3-preview/acb5b1b0522237f021af432207d61ca0fca28d48?cid=8897482848704f2a8f8d7c79726a70d4",
      "track_number" : 4,
      "type" : "track",
      "uri" : "spotify:track:5YXr4AGfUQpLSxtFSsKUh6"
    }, {
      "artists" : [ {
        "external_urls" : {
          "spotify" : "https://open.spotify.com/artist/12Chz98pHFMPJEknJQMWvI"
        },
        "href" : "https://api.spotify.com/v1/artists/12Chz98pHFMPJEknJQMWvI",
        "id" : "12Chz98pHFMPJEknJQMWvI",
        "name" : "Muse",
        "type" : "artist",
        "uri" : "spotify:artist:12Chz98pHFMPJEknJQMWvI"
      } ],
      "available_markets" : [ "AD", "AR", "AT", "AU", "BE", "BG", "BO", "BR", "CA", "CH", "CL", "CO", "CR", "CY", "CZ", "DE", "DK", "DO", "EC", "EE", "ES", "FI", "FR", "GB", "GR", "GT", "HK", "HN", "HU", "ID", "IE", "IS", "IT", "JP", "LI", "LT", "LU", "LV", "MC", "MT", "MX", "MY", "NI", "NL", "NO", "NZ", "PA", "PE", "PH", "PL", "PT", "PY", "SE", "SG", "SK", "SV", "TH", "TR", "TW", "US", "UY", "VN" ],
      "disc_number" : 1,

HISTORY.md
# v0.4.0 - 2017.10.03

* Added support for album retrieval

# v0.3.6 - 2016.10.12

* Added support for `market` when searching

# v0.3.5 - 2016.05.12

* Removed extraneous diagnostic code

# v0.3.4 - 2016.05.12

* Relaxed sanitization routined used to clean up search terms prior to sending to the Spotify API (no longer filtering out `(` and `)` characters)

# v0.3.3 - 2015.01.14

* Relaxed sanitization routine used to clean up search terms prior to sending to the Spotify API (no longer filtering out `*` characters)

# v0.3.2 - 2015.12.11

* Added support for returning multiple closest matches for when multiple Spotify tracks match the same ISRC or the same artist and title

# v0.3.1 - 2015.11.04

* Reworked how errors are returned when captured from Spotify API

# v0.3.0 - 2015.11.04

* Breaking interface change - search and findBestMatch accept `params` object now
* Added support for lookup based on ISRC code
* Added support for [Spotify Web API Client Credentials ](https://developer.spotify.com/web-api/authorization-guide/#client_credentials_flow)

# v0.2.1 - 2014.12.03

* Updating sanitize internal function to filter less - fixes defect where occaisionally title and artist was blank

# v0.2.0 - 2014.10.22

* Simplify response from search
* Added support for search result limit
* Added support for start index (rather than page) on search
* Changed query interface to support artist and title explicitly
* Updated to use newer version of Spotify search API

# v0.1.1 - 2014.10.22

* Renamed internal constructor to reflect module name
* Updated dependencies
* Add repository for package.json file

# v0.1.0 - 2014.05.22

* Initial release


lib/index.js
/* jshint latedef: nofunc */
var
	levenshtein = require('levenshtein'),
	request = require('request'),
	unidecode = require('unidecode');


var spotifier = (function (self) {
	'use strict';

	self = self || {};

	var
		accessToken,
		defaultOptions = {
			albumUrl : 'https://api.spotify.com/v1/albums',
			authorizationUrl : 'https://accounts.spotify.com/api/token',
			searchResultLimit : 10,
			searchUrl : 'https://api.spotify.com/v1/search',
			timeout : 10000 // 10 seconds
		};

	function closestMatch (artist, title, matches) {
		var
			artistName = '',
			bestDistance = null,
			bestMatches = [],
			distance = 0,
			distances = [];

		// sanitize query
		artist = sanitize(artist);
		title = sanitize(title);

		// iterate each match and determine the distance from the normalized
		// search artist and title
		matches.forEach(function (match) {
			artistName = (match.artists && match.artists.length) ?
				match.artists[0].name :
				'';

			// jshint newcap: false
			distance = new levenshtein(
				artist + ' ' + title,
				sanitize(artistName) + ' ' + sanitize(match.name))
			.distance;

			// store the distance
			distances.push(distance);

			if (bestDistance === null || distance < bestDistance) {
				bestDistance = distance;
			}
		});

		// retrieve all matches that tie with the best match
		distances.forEach(function (distance, i) {
			if (distance === bestDistance) {
				bestMatches.push(matches[i]);
			}
		});

		return bestMatches;
	}

	function ensureAccessToken (authorization, callback) {
		// if there is nothing to authenticate with, return immediately
		if (!authorization || !authorization.length) {
			return setImmediate(callback);
		}

		// if we already have a valid access token, return it
		if (accessToken && new Date() < accessToken.expiresAt) {
			return setImmediate(function () {
				return callback(null, accessToken);
			});
		}

		var requestedAt = new Date();

		return request({
			form : 'grant_type=client_credentials',
			headers : {
				Authorization : ['Basic', authorization].join(' ')
			},
			method : 'POST',
			url : self.options.authorizationUrl
		}, function (err, res, body) {
			if (err) {
				return callback(err);
			}

			if (res.statusCode >= 200 && res.statusCode <= 299) {
				if (body) {
					try {
						accessToken = JSON.parse(body);
					} catch (ex) {
						ex.body = body;
						return callback(ex);
					}

					// jshint sub : true
					accessToken.expiresAt = new Date(
						Number(requestedAt) + (accessToken['expires_in'] * 1000));

					return callback(null, accessToken);
				}
			}

			return callback(new Error(body));
		});
	}

	function sanitize (term) {
		var
			i = 0,
			reFeaturing = /\sfeat(uring)?/g, // match " feat" and " featuring"
			reNonAllowableChar = /[^\*\.\(\)\-_'a-zA-Z 0-9]+/g; // match anything that is not alpha, numeric, space, -, _ and '

		term = unidecode(term.toLowerCase())
			.replace(reNonAllowableChar, ' ') // remove unallowed chars !
			.split(/\s+/).join(' '); // normalize      spaces

		// remove featuring
		i = term.search(reFeaturing);
		if (i > 0) {
			term = term.substring(0, i);
		}

		return term;
	}

	self.findBestMatch = function (options, params, callback) {
		if (typeof callback === 'undefined' && typeof params === 'function') {
			callback = params;
			params = options;
			options = {};
		}

		var bestMatches = [];

		return self.search(options, params, function (err, result) {
			if (err) {
				return callback(err);
			}

			if (result.total > 0) {
				if (params.isrc) {
					bestMatches = result.items;
				} else {
					bestMatches = closestMatch(params.artist, params.title, result.items);
				}
			}

			// if the multi param is supplied, return all top matches
			return callback(null, options.multi ? bestMatches : bestMatches[0]);
		});
	};

	self.lookupAlbum = function (options, albumId, callback) {
		if (typeof callback === 'undefined' && typeof albumId === 'function') {
			callback = albumId;
			albumId = options;
			options = {};
		}

		var
			authorization = '',
			clientId = options.clientId || self.options.clientId,
			clientSecret = options.clientSecret || self.options.clientSecret,
			result = {};

		if (!albumId) {
			return setImmediate(
				callback,
				new Error('albumId must supplied'));
		}

		// check for authorization
		if (clientId && clientSecret) {
			authorization = new Buffer(
				[clientId, clientSecret].join(':')).toString('base64');
		}

		// ensure an access token (if applicable)
		return ensureAccessToken(authorization, function (err, token) {
			if (err) {
				return callback(err);
			}

			// jshint sub : true
			return request({
				headers : {
					Authorization : (token ?
						['Bearer', token['access_token']].join(' ') :
						undefined)
				},
				method : 'GET',
				timeout : self.options.timeout,
				uri : [self.options.albumUrl, albumId].join('/')
			}, function (err, res, body) {
				if (err) {
					return callback(err);
				}

				if (body) {
					try {
						result = JSON.parse(body);
					} catch (ex) {
						ex.body = body;
						return callback(ex);
					}
				}

				// if the response was a success, return the body
				if (res.statusCode >= 200 && res.statusCode <= 299) {
					return callback(null, result || {});
				}

				// return the error response
				return callback(result);
			});
		});
	};

	self.search = function (options, params, callback) {
		if (typeof callback === 'undefined' && typeof params === 'function') {
			callback = params;
			params = options;
			options = {};
		}

		var
			authorization = '',
			clientId = options.clientId || self.options.clientId,
			clientSecret = options.clientSecret || self.options.clientSecret,
			market = options.market || params.market || self.options.market,
			query = '',
			result = {};

		if (params.isrc) {
			query = ['isrc:', params.isrc].join('');
		}

		// sanitize artist and title if supplied
		if (params.artist && params.title) {
			query = [
				'artist:',
				sanitize(params.artist),
				'+track:',
				sanitize(params.title)].join('');
		}

		if (!query.length) {
			return setImmediate(
				callback,
				new Error('either isrc or artst and title must be supplied as params'));
		}

		// check for authorization
		if (clientId && clientSecret) {
			authorization = new Buffer(
				[clientId, clientSecret].join(':')).toString('base64');
		}

		// ensure an access token (if applicable)
		return ensureAccessToken(authorization, function (err, token) {
			if (err) {
				return callback(err);
			}

			// jshint sub : true
			return request({
				headers : {
					Authorization : (token ?
						['Bearer', token['access_token']].join(' ') :
						undefined)
				},
				method : 'GET',
				qs : {
					limit : options.limit || self.options.searchResultLimit,
					market : market,
					offset : options.start || 0,
					type : 'track',
					useQuerystring : true
				},
				timeout : self.options.timeout,
				uri : [self.options.searchUrl, '?q=', query].join('')
			}, function (err, res, body) {
				if (err) {
					return callback(err);
				}

				/*
				// debugging
				console.log([
					self.options.searchUrl,
					'?q=',
					query,
					'&limit=',
					options.limit || self.options.searchResultLimit,
					'&offset=',
					options.start || 0,
					'&type=track',
					'&market=',
					market].join(''));
				//*/

				if (body) {
					try {
						result = JSON.parse(body);
					} catch (ex) {
						ex.body = body;
						return callback(ex);
					}
				}

				// if the response was a success, return the body
				if (res.statusCode >= 200 && res.statusCode <= 299) {
					return callback(null, result.tracks || []);
				}

				// return the error response
				return callback(result);
			});
		});
	};

	return function (options) {
		options = options || {};

		// apply default for missing keys
		Object.keys(defaultOptions).forEach(function (key) {
			if (typeof options[key] === 'undefined') {
				options[key] = defaultOptions[key];
			}
		});

		self.options = options;

		return self;
	};
}({}));

exports = module.exports = spotifier;
exports.initialize = spotifier;
